src/LibSignedSettlement.sol

```diff
@@ -0,0 +1,376 @@
+                v := and(calldataload(add(lastByte, 65)), 0xff)
+            }
+        }
+    }
+
+    /// @dev Deadline, signature and offsets are encoded at the end of the calldata. It is not just used directly
+    ///      as a parameter to the function because that'd lead to the deadline getting
+    ///      encoded inplace and messing up all the offets which'd mean we cannot directly
+    ///      copy the calldata as-is for the call to the actual settlement contract.
+    ///
+    ///      Encoding for the appended data is expected to be `abi.encodePacked(deadline, r, s, v, offsets)`.
+    ///      `v` is assumed to be uint8, while all others will encode to full 32 byte word.
+    ///      `offsets` is assumed to be `uint[3]`.
+    ///
+    ///      Instead of reading the last 97 bytes directly with `calldataload`
+    ///      we determine the expected byteoffset to read based on the last post interaction. Acts as a
+    ///      a validation that a deadline was infact encoded at the end of calldata and not accidentally
+    ///      reading a word from the last interaction's encoded data.
+    function readExtraParamsPartiallySigned(
+        address[] calldata tokens,
+        uint256[] calldata clearingPrices,
+        GPv2Trade.Data[] calldata trades,
+        GPv2Interaction.Data[][3] calldata interactions
+    )
+        internal
+        pure
+        returns (uint256 deadline, uint256 r, uint256 s, uint256 v, uint256[3] calldata offsets, uint256 lastByte)
+    {
+        {
+            bytes calldata extraBytes;
+            (extraBytes, lastByte) = readExtraBytes(tokens, clearingPrices, trades, interactions);
+
+            if (lastByte != msg.data.length - 193) {
+                revert LibSignedSettlement__InvalidExtraParamsPartiallySigned();
+            }
+
+            assembly ("memory-safe") {
+                deadline := calldataload(lastByte)
+                r := calldataload(add(lastByte, 0x20))
+                s := calldataload(add(lastByte, 0x40))
+                v := and(calldataload(add(lastByte, 0x41)), 0xff)
+                offsets := add(lastByte, 0x61)
+            }
+
+            // validate that the subset size is <= interactions length
+            if (
+                offsets[0] > interactions[0].length || offsets[1] > interactions[1].length
+                    || offsets[2] > interactions[2].length
+            ) {
+                revert LibSignedSettlement__InvalidExtraParamsPartiallySigned();
+            }
+        }
+    }
+
+    /// @dev Reads params appended at the end, computes the digest for the fully signed message
+    ///      and returns the memory slice where the calldata is encoded for the `GPv2Settlement.settle`
+    ///      call.
+    ///
+    ///      The digest is assumed to be keccak256 hash of the following encoded data:
+    ///      `abi.encodePacked(abi.encode(tokens, clearingPrices, trades, interactions), abi.encode(deadline, solver))`
```

Comment:
If we encode the solver, shouldn't there be some check somewhere that `msg.sender` == solver? I cannot find that assertion. Is there a test case for that?

Reply:
yes, @fedgiac is correct, its checked by using caller